Metalama.Framework.Engine/Advising/IntroduceConstructorAdvice.cs

```diff
@@ -0,0 +1,151 @@
+﻿// Copyright (c) SharpCrafters s.r.o. See the LICENSE.md file in the root directory of this repository root for details.
+
+using Metalama.Framework.Advising;
+using Metalama.Framework.Aspects;
+using Metalama.Framework.Code;
+using Metalama.Framework.Code.DeclarationBuilders;
+using Metalama.Framework.Engine.Aspects;
+using Metalama.Framework.Engine.CodeModel;
+using Metalama.Framework.Engine.CodeModel.Builders;
+using Metalama.Framework.Engine.CodeModel.References;
+using Metalama.Framework.Engine.Diagnostics;
+using Metalama.Framework.Engine.Services;
+using Metalama.Framework.Engine.Transformations;
+using Metalama.Framework.Engine.Utilities;
+using System;
+using System.Collections.Generic;
+
+namespace Metalama.Framework.Engine.Advising;
+
+internal sealed class IntroduceConstructorAdvice : IntroduceMemberAdvice<IMethod, IConstructor, ConstructorBuilder>
+{
+    private readonly PartiallyBoundTemplateMethod _template;
+
+    private new Ref<INamedType> TargetDeclaration => base.TargetDeclaration.As<INamedType>();
+
+    public IntroduceConstructorAdvice(
+        IAspectInstanceInternal aspect,
+        TemplateClassInstance templateInstance,
+        INamedType targetDeclaration,
+        ICompilation sourceCompilation,
+        PartiallyBoundTemplateMethod template,
+        IntroductionScope scope,
+        OverrideStrategy overrideStrategy,
+        Action<IConstructorBuilder>? buildAction,
+        string? layerName,
+        IObjectReader tags )
+        : base(
+            aspect,
+            templateInstance,
+            targetDeclaration,
+            sourceCompilation,
+            null,
+            template.TemplateMember,
+            scope,
+            overrideStrategy,
+            buildAction,
+            layerName,
+            tags )
+    {
+        this._template = template;
+
+        this.Builder = new ConstructorBuilder(this, targetDeclaration );
+    }
+
+    protected override void InitializeCore(
+        ProjectServiceProvider serviceProvider,
+        IDiagnosticAdder diagnosticAdder,
+        TemplateAttributeProperties? templateAttributeProperties )
+    {
+        var targetDeclaration = this.TargetDeclaration.GetTarget( this.SourceCompilation );
+
+        switch ( this.OverrideStrategy )
+        {
+            case OverrideStrategy.New:
+                diagnosticAdder.Report(
+                    AdviceDiagnosticDescriptors.CannotUseNewOverrideStrategyWithFinalizers.CreateRoslynDiagnostic(
```

Comment:
I think the diagnostic here needs to be adjusted. It shouldn't talk about finalizers when introducing a constructor.